app/src/components/report/LocationPicker.tsx

```diff
@@ -0,0 +1,222 @@
+    useControl(() => geocoder);
+
+    geocoder.on('result', (event) => {
+        const [longitude, latitude] = event.result.geometry.coordinates;
+        placeMarker({ longitude, latitude });
+    });
+
+    return null;
+}
+
+/** The map */
+export default function LocationPicker({
+    reportCoords,
+    updateReportCoords,
+    fillOutAddress,
+    initialMarkerState,
+    locationSubmitted,
+}: LocationPickerProps) {
+    /** Retrieve last user reported location for the view of the map upon loading page*/
+    function fetchUserLastLocation(): ViewStateType {
+        if (typeof window === 'undefined') return DEFAULT_COORDINATES;
+
+        if (isUndefined(reportCoords)) {
+            const lastReportedLocation = localStorage.getItem('lastReportedLocation');
+
+            if (lastReportedLocation) {
+                const { longitude, latitude } = JSON.parse(lastReportedLocation);
+                return { longitude, latitude, zoom: 10.5 };
+            } else {
+                return DEFAULT_COORDINATES;
+            }
+        } else {
+            return { ...reportCoords!, zoom: 10.5 };
+        }
+    }
+    // View Box of the map upon entering a page
+    const [viewState, setViewState] = useState<ViewStateType | {}>(fetchUserLastLocation);
+    // Disable the marker after user already submit location
+    const [draggableMarker, setDraggableMarker] = useState<boolean>(initialMarkerState);
+    // Ref to open modal alert
+    const modalRef = useRef<ModalHandle | null>(null);
+
+    function handleDragEnd(event: MarkerDragEvent) {
+        const { lng, lat } = event.lngLat;
+        updateReportCoords({ longitude: lng, latitude: lat });
+    }
+
+    // Track user location and put the marker there only if the location is not submitted
+    function handleGeolocate(event: GeolocateResultEvent) {
+        if (!draggableMarker) return;
+        const { longitude, latitude } = event.coords;
+        updateReportCoords({ longitude, latitude });
+    }
+
+    // Reset current location and get new location
+    function resetLocationAndAddress() {
+        setDraggableMarker(true);
+        fillOutAddress('');
+    }
+
+    // Translate coordinates to address
+    async function reverseGeolocation() {
+        if (isUndefined(reportCoords)) return;
+
+        const { longitude, latitude } = reportCoords!;
+
+        if (!longitude || !latitude) return;
+
+        const endpoint = 'mapbox.places';
+        try {
+            const response = await fetch(
+                `https://api.mapbox.com/geocoding/v5/${endpoint}/${longitude},${latitude}.json?types=address,poi,place&access_token=${MAPBOX_ACCESS_TOKEN}`
+            );
+            const json = await response.json();
+            const results = json.features[0];
+            fillOutAddress(results.place_name);
+            // Disable marker after submitted location
+            setDraggableMarker(false);
+        } catch (error) {
+            console.log(error);
+        }
+    }
```

Comment:
Non-blocking / up to you:

This seems like the sort of logic that would be helpful to abstract into a `MapBoxClient` with a function `reverseGeolocate(longitude: number, latitude: number): string`

That way, we can use this logic in other places in our application easily without having to copy-and-paste :)

Also, if we have an interface `IMapClient` with an implementation `MapBoxClient`, then we could make other `IMapClient` implementations with other data providers (AWS, Google Maps, etc.)